### DISCUSSION TITLE
Best Practices Error Handling

### DISCUSSION THREAD

**Discussion by joa23** [danielgtaylor/huma#424] [Q&A] [answered]
Hi Community, 
What are people doing regarding good but simple error handling in Huma?
I have handler -> service -> sqlc / DB layer. I don't want to expose the details of the error to the end user but adding own errors with error codes in each layer seems like a lot of unnecessary work. 
One requirement we have is that we attach the error on certain locations of our json objects to handle them in the UI appropriately. 
For example: userName <- "username already exists" 
Thanks for your suggestions.

**Comment by danielgtaylor** [comment, marked as answer]
I would typically handle this at the operation handler, converting as you need after bubbling up any of the internal errors without translation. It's a bit of extra work but makes the code very easy to reason about / maintain and explicitly converts between the layers. Something like:

```go
func MyHandler(ctx context.Context, input *MyInput) (*MyOutput, error) {
  result, err := db.Get("some query")
  if errors.Is(err, db.ErrorNotFound) {
    slog.ErrorContext(ctx, "Not found", "err", err, "id", input.ID)
    return nil, huma.Error404NotFound("Thing with ID " + input.ID + " not found")
  }

  if errors.Is(err, db.ConnectionReset) {
    slog.ErrorContext(ctx, "Can't talk to the DB", "err", err, "id", input.ID)
    return nil, huma.Error500InternalServerError("Unable to reach DB, please try again later")
  }

  // Otherwise it's the happy path
  return &MyOutput{Body: /* ... */}, nil
}
```

**Reply by joa23** [reply]
Thank you so much.
